.github/actions/sonar/action.yml

```
@@ -0,0 +1,253 @@
+            echo "Tasks are still in progress or pending. Waiting..."
+            sleep 10
+          else
+            echo "All tasks completed."
+            break
+          fi
+        done' || (echo "SonarQube tasks failed to complete in time" && exit 1)
+
+    # Retrieve SonarQube metrics for the PR
+    - name: 'Retrieve SonarQube metrics'
+      if: github.event_name == 'pull_request' && steps.restore_cache.outputs.cache-hit == 'true'
+      id: sonar_metrics
+      shell: bash
+      run: |
+        SONAR_RESPONSE=$(curl -s -u admin:admin \
+        "http://localhost:9000/api/measures/component?component=${{ inputs.sonar-project-key }}&pullRequest=${{ github.event.pull_request.number }}&metricKeys=new_bugs,new_vulnerabilities,new_code_smells,new_coverage,new_duplicated_lines_density,new_violations")
+
+        export_measure() {
+          METRIC_NAME=$1
+          ENV_VAR_NAME=$2
+          METRIC_VALUE=$(echo "$SONAR_RESPONSE" | jq -r ".component.measures[] | select(.metric == \"$METRIC_NAME\") | .period.value")
+          if [ -z "$METRIC_VALUE" ] || [ "$METRIC_VALUE" == "null" ]; then
+            METRIC_VALUE="N/A"
+          fi
+          echo "$ENV_VAR_NAME=$METRIC_VALUE" >> $GITHUB_ENV
+          export $ENV_VAR_NAME=$METRIC_VALUE
+        }
+
+        export_measure "new_vulnerabilities" "NEW_VUL"
+        export_measure "new_code_smells" "NEW_CSM"
+        export_measure "new_bugs" "NEW_BUG"
+        export_measure "new_violations" "NEW_VIOLATIONS"
+        export_measure "new_coverage" "NEW_COV"
+        export_measure "new_duplicated_lines_density" "NEW_DUP"
+
+        {
+          echo "## :bar_chart: SonarQube Analysis for ${{ inputs.sonar-project-key }}"
+          echo ""
+          echo "| Metric                      | Value       |"
+          echo "|-----------------------------|-------------|"
+          echo "| **New Vulnerabilities**     | ${NEW_VUL} |"
+          echo "| **New Bugs**                | ${NEW_BUG} |"
+          echo "| **New Code smells**         | ${NEW_CSM} |"
+          echo "| **Coverage on New Code**    | ${NEW_COV}% |"
+          echo "| **Duplication on New Code** | ${NEW_DUP}% |"
+
+          if [[ "${NEW_VIOLATIONS}" != "N/A" && "${NEW_VIOLATIONS}" != "0" ]]; then
+            echo ""
+            echo "<details><summary>Unresolved Issues (click to expand)</summary>"
+            echo ""
+            ISSUES=$(curl -s -u admin:admin \
+              "http://localhost:9000/api/issues/search?componentKeys=${{ inputs.sonar-project-key }}&resolved=false&pullRequest=${{ github.event.pull_request.number }}" | \
+              jq -r '.issues[] | "File: \(.component) Line: \(.line)\n  [\(.rule)] \(.message)\n"')
+            echo "$ISSUES"
+            echo "</details>"
+          fi
+        } > sonar_result.md
+
+        {
+          echo 'COMMENT_BODY<<EOF'
+          cat sonar_result.md
+          echo 'EOF'
+        } >> "$GITHUB_ENV"
```

Print result to log.
```suggestion
        } >> "$GITHUB_ENV"

        cat sonar_result.md
```